ihontaryk/homework_11/mine_sweeper.py

```diff
@@ -0,0 +1,186 @@
+        """
+
+        data = tuple(map(int, data.strip().split()))
+        assert len(data) == 2
+        assert data in self.empty_cells() + self.mines
+        return data
+
+    def empty_cells(self) -> list:
+        return [(x, y)
+                for x in list(range(self.width))
+                for y in list(range(self.height))
+                if (x, y) not in self.mines
+                and (x, y) not in self.guesses]
+
+    def in_field(self, coord: tuple) -> bool:
+        """
+        Check that coord is in field
+        """
+
+        return coord in self.empty_cells()
+
+    def is_mine(self, coord: tuple) -> bool:
+        """
+        Check if there is a mine under coord
+        """
+
+        return coord in self.mines
+
+    def mines_around(self, coord: tuple) -> int:
+        """
+        Calculate, how many mines are in neighbouring cells
+        """
+
+        x, y = coord
+        return sum((x + dx, y + dy) in self.mines
+                   for (dx, dy) in self.NEIGHBOURS)
+
+    def save_guess(self, coord: tuple):
+        """
+        Save user input,
+        in case if cell is 0, recursively open neighbour cells
+        """
+
+        queue = [coord]
+        while queue:
+            x, y = queue.pop()
+            mines_around = self.mines_around((x, y))
+            self.guesses[(x, y)] = mines_around
+
+            # recursively open all neighbour cells if current cell == 0
+            if mines_around == 0:
+                neighbour_cells = {(dx + x, dy + y) for dx, dy in self.NEIGHBOURS}
+                cells_to_check = neighbour_cells.intersection(self.empty_cells())
+                queue.extend(cells_to_check)
+
+    def win(self):
+        """
+        Check if user wins the game
+        """
+
+        return not self.empty_cells()
+
+    def main(self):
+        while not self.win():
+            # show screen with current state
+            self.screen.update_screen()
+            # input from user
+            data = self.receive_user_input()
+
+            # validate and normalize
+            # if error - catch, inform user and return to start
+            try:
+                data = self.normalize_user_input(data)
+            except Exception:
+                self.screen.show_alert("Wrong data: {}".format(data))
+                continue
+
+            # if mine - end game
+            if self.is_mine(data):
+                self.screen.show_alert("Is a mine, you loose: {}".format(data))
+                break
+
+            # save user input
+            self.save_guess(data)
+
+        # if win show win message
+        else:
+            self.screen.show_alert("You Win")
+
+        # show screen last time
+        self.screen.update_screen(show_mines=True)
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    height, width, mine_count = CONFIG.get(args.difficulty) or CONFIG["hard"]
+    GameSession(height, width, mine_count).main()
```

Comment:
pass config as 1 single param, since it is a set of dependant params
it is common approach to reduce amount of input params but not to increase